### ISSUE TITLE
transmission-daemon extreemly high memory usage

### ISSUE THREAD

**Issue by Mechazawa** [transmission/transmission#313]
I currently have five sessions running and they take up all availible memory. 

```
$ free -h
              total        used        free      shared  buff/cache   available
Mem:            15G         15G        157M         11M        127M         50M
Swap:          8.0G        4.8G        3.2G
```

``` 
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root     31717  0.4  4.0 5475364 662052 ?      Ssl  Jun18  20:47 /usr/bin/transmission-daemon -f -g /path/to/config
root     31619  0.3  4.4 5017080 719940 ?      Ssl  Jun18  16:25 /usr/bin/transmission-daemon -f -g /path/to/config
root     31501  0.8  6.0 5278224 990924 ?      Ssl  Jun18  38:28 /usr/bin/transmission-daemon -f -g /path/to/config
root     31859  1.0  9.3 8096668 1530164 ?     Ssl  01:38   6:08 /usr/bin/transmission-daemon -f -g /path/to/config
root      4456  4.5 67.0 15763784 10955100 ?   Ssl  04:57  16:30 /usr/bin/transmission-daemon -f -g /path/to/config
```

**Comment by messyidea** [comment]
Maybe it it a bug of debian 9.
Today I have meet this problem too.
I have build a staticlly linked transmission-daemon. It works ok.

Here is a temporary way to solve this problem.
Replace the /usr/bin/transmission-daemon with staticlly linked transmission.
Change the "/etc/systemd/system/multi-user.target.wants//etc/systemd/system/multi-user.target.wants" as follows:
```
[Unit]
Description=Transmission BitTorrent Daemon
After=network.target

[Service]
User=debian-transmission
Type=simple
ExecStart=/usr/bin/transmission-daemon -f --log-error
ExecStop=/bin/kill -s STOP $MAINPID
ExecReload=/bin/kill -s HUP $MAINPID


[Install]
WantedBy=multi-user.target
```
Then run the command:
```
systemctl daemon-reload
systemctl restart transmission-daemon
```
[transmission-daemon.zip](https://github.com/transmission/transmission/files/1097986/transmission-daemon.zip)

**Comment by Mechazawa** [comment]
Recompiling transmission with static linking seems to have done the trick

**Comment by Astara** [comment]
recompiling transmission w/static linking will cause transmission to keep separate copies of system libraries in memory for itself -- using more memory.

I only have about 350 torrents using 7.5G virtual, 2g resident, since each torrent is many GB long, I would hope it uses alot of memory for buffering.  According to top, I see transmission using about 2.092% of memory and about 1.3% of of one cpu.  The more memory I give it for buffering, the lower the cpu usage and less disk activity.

How big are the torrents you are serving?  A single suse release comes out on DVD and could easily take >4G.   You want to be able to specify memory usage, since if you set it too low, it will slow down torrent serving and raise unnecessary disk activity.

**Comment by mikedld** [comment]
@messyidea, @Mechazawa, whey you guys talk about "static linking", what exactly do you mean by that? What are the dependencies of transmission-daemon (as reported by ldd) provided by the distro versus the one you built? It seems unlikely to me that just rebuilding something *with same set of dependencies* will produce different results, no matter how the linking was performed... E.g. could it be that when rebuilding you configured Transmission with uTP disabled while it was enabled in the distro package or vice versa? Or used different crypto backend (openssl, wolfssl, mbedtls)?

**Comment by messyidea** [comment]
@mikedld 
Here is my build script.
https://gist.github.com/messyidea/16f9c63d7b219d2a4755e30aedd8268e
Based on this repo.
https://github.com/lancethepants/transmission-mipsel-static/blob/master/transmission.sh

**Comment by Mechazawa** [comment]
I tried a regular rebuild from sources yesterday and that seems to be working fine as well for me.

**Comment by xavery** [comment]
Same thing happened to me while running the Transmission build provided by Debian buster (testing), which is also 2.92 like Debian 9. Running about 200 torrents, mostly seeding and idle. Memory usage as reported by htop would skyrocket to 1088M after about 5 minutes of usage even if the torrents weren't too active during that period.
Compiling the current HEAD at the time of writing (6410a460ea3acc83a3573f70acf6687d12879d76) seems to have fixed the issue, memory usage is now around 60-80M.

**Comment by xavery** [comment]
I got hit by this issue as well. I set up a simple memory usage monitor which pulls the resident set size from /proc and asks the daemon for the current download/upload speeds, as reported to systemd.
I was hoping that the download/upload speeds would correlate with the memory usage. However, it seems like the daemon uses a ton of memory at start (the machine was swapping a lot), and then - somehow - the memory usage stabilises to around 800MB and stays this way irrelevant of the down/up speeds. I have about 200 running torrents in the session.
You can see the graph [here](http://i.imgur.com/faymneP.png). The x axis represents the number of seconds since the monitoring was started. The memory usage is in pages, multiply by 4 to get KBs. I can provide the data file on request.

Now, I'm no Transmission expert by any means, but I guess this cannot be related to the `cache-size-mb` setting at all : there is only one cache object for the whole Transmission session, which is resized only while reading the settings to make its size equal to the one requested in the settings. Then I thought that maybe I/O prefetching is responsible for this situation (since my Transmission was compiled with support for `posix_fadvise`), but then the memory usage wouldn't be counted towards the daemon process itself, but rather towards the kernel's I/O buffers reported as the `buff/cache` column in the output of `free`.

**Comment by nakhan98** [comment]
I'm experiencing the same issue with memory leaks on 2.92 on raspbian stretch (raspberry pi model b+ w/ 512mb ram) . The version in the raspbian repository is unusable and immediately jumps to 50% memory usage after startup according to htop and the system soon after starts swapping and becomes unresponsive.

Building 2.92 from source works better but memory usage is still high (around 30%) compared to 2.84. And memory leaks occur if i seed from a network mounted drive (sshfs) .

I only upgraded to stretch a few weeks ago and prior to this I had no such problems with, 2.84 on raspbian jessie. It was rock solid and transmission was running constantly for months at a time with zero issues.

I built and installed 2.84 from source yesterday everything seems back to normal so it's unlikely to be an OS issue. Memory usage rarely exceeds 20% and at the moment is hovering under 15%

Other relevant info, I have cache-size-mb set to 8. I have around 700 torrents loaded but about 250 are seeding at any given time.

**Comment by Astara** [comment]
I'm looking at 2 of my torrents -- one 400MB torrent has a piece size of 256KB, another 6G torrent has a piece size of 8M.  Of the next 6, 3 have 8MB piece sizes w/the lowest size at 512KB and 4 have 4MB piece sizes.  That's 1 piece of the torrent that has to take 4MB of memory while it is sending 1 piece to 1 client.  If all of the torrents you serve have a piece size of 1MB each, then to have 256 of them sending or receiving 1 piece to 1 client at a time would take 256MB of buffer memory.  

If 2.84 was working "rock solid" for you, I'd stick with it.

**Comment by andreygursky** [comment]
Once I've been forced to change the announcer URLs from http to https, transmission-daemon (2.92 trunk 14736) a couple of minutes after start deadlocks Debian 9 testing (Buster) running on 1.5G available RAM (without swap) for 30-60 minutes (no chance to kill manually) until oom-killer does the job. I've figured out a workaround: now I must start such torrents in small batches and delayed for several minutes inbetween, since transmission-daemon doubles its memory consumption after announcing just about 50 torrents.

**Comment by Astara** [comment]
Having an announcer URL be https would be very resource intensive.  It's the setting up of the TCP connection that is most intense (which was why UDP was added for announcer URL's). Changing it to HTTPS slows it down by several factors.  I wouldn't think it would be reasonable to try to support announcing in https -- I just wouldn't think it would be technically feasible with today's hardware.

**Comment by andreygursky** [comment]
My workaround just delayed the issue a little bit. I've seen spikes in the memory consumption until it hanged again. It seems, that if announcing fails for some reason, than more and more torrents get announced at the same time, which I have to avoid. So digging into sources is unavoidable.

@Astra, yes, I'm aware that TCP with SSL costs more than clear-text UDP, but I'm not speaking of millions of torrents. 1.5GB of RAM are not enough for 50 TCP connections with SSL? This must be a bug in transmission or in a way, how transmission uses libcurl, I guess.

**Comment by Mechazawa** [comment]
I refuse to believe that it has anything to do with SSL. It feels a lot more like a memory leak.

**Comment by andreygursky** [comment]
@Mechazawa, until I switched announcers to SSL, I had never problems. And a memory leak cannot fluctuate, it can only accumulate, instead I'm observing spikes in memory consumptions, which, I guess, coincide with reannouncements.

**Comment by Mechazawa** [comment]
@andreygursky I'm currently graphing the memory usage of one of the daemons. I'll post the results once it crashes due to lack of memory

**Comment by andreygursky** [comment]
@Mechazawa, what favor of curl development package was installed when you recompiled transmission (GnuTLS, OpenSSL, NSS,..)?

**Comment by Astara** [comment]
@Mechazawa You've already said that recompiling with static linking fixed the problem.  That proves it can' be a problem in the code and can't be a memory leak.  Besides, what does a memory leak that goes away w/static leaking feel like?

If you believe it to be a memory leak -- show that it is so.  Run it under Valgrind.  It can track every allocation.  Compile transmission w/symbols and it can point exactly to what memory is "lost" (no references to it) or what hasn't been freed at exit (but may still be referenced).  Most import is memory that has no references to it at program exit.  On top of that, valgrind will show you exactly where the memory that was leaked was allocated.  

I've had transmission 2.92 running for ***months***, and not had it change in its memory usage -- I don't use SSL but I do allow encryption when requested.  None of the sites I connect to use SSL trackers -- mostly http, some udp and some magnets.  But no way can I believe transmission can run for months, have a memory leak problem and NOT crash or run out of memory.  I've been restarting it more recently to free up network bandwidth, but my main daemon (I usually run 2 instances to get better priority grading) has been running since Sep 12... this is the 27'th.. that's 15 days so far -- not that I'd expect it to be any different, as the binary hasn't changed.

**IF** there's a memory leak, it has to be in code I don't use -- like the SSL announce code -- which is a real bad idea anyway.  Looking around the web I see figures of 200-400ms/connection -- that would be repeated every announce and with every torrent.  50 torrents might take as much as 25 seconds.  If running xmission in a container might take double or triple that.  

That reminds me -- @andreygursky -- how many cpu's does your machine have and are you running transmission in a container or VM?  Networking would be one of the slower virtualization areas for connection creation/teardown.  If you are doing SSL in a container, 50 torrents **might** tax things (I wouldn't think it likely), but costs of repeated SSL session creation/teardown times virtualization costs might add up...

BTW -- I'm not a transmission developer, but I do build my own and have used it for several years.  I usually have it running 24x7 on my home machine, so I'd notice it if it was misbehaving...

**Comment by andreygursky** [comment]
I've removed libcurl4-gnutls-dev and libssl-dev and installed libcurl4-openssl-dev and libssl1.0-dev instead. Then I've rebuilt transmission. This seems to solve the issue. Searching reveals following: https://stackoverflow.com/questions/45498537/https-request-memory-leak-with-curlasynchttpclient and https://github.com/curl/curl/issues/1086. Don't know yet how related.

If my issue is unrelated to the original one, let me know, to open a new one or @mikedld could just add a new entry into known issues list?

P.S. It's a little bit strange, Debian cherry-picked OpenSSL 1.1 patch, but uses GnutTLS flavour of libcurl. And even if you like to use OpenSSL flavour of libcurl, it is only of version 1.0, while mixing is not possible (and not something good in general), thus using OpenSSL 1.1 is not possible at the moment.

**Comment by andreygursky** [comment]
@Astara, it's a pretty old PC with one core 64bit CPU.

**Comment by Mechazawa** [comment]
Running 4 daemons with ±2000 torrents each (all of them have ssl announce urls). I've graphed out memory usage for now (till the daemon crashed). 

![image](https://user-images.githubusercontent.com/1144171/30958156-cea72722-a42b-11e7-8937-7f28ee75a133.png)


I don't have time right now to look at it more in-depth though.

**Comment by Astara** [comment]
2000 torrents each?   I use between 1.5-3Gb with 350 torrents with no SSL.  Is that the memory usage for all 4 daemons?  So about 8000 torrents?  Have you ever had this working w/transmission or any other torrent program?  Just seems like quite a few torrents.  If it is a memory leak, you should be able to reproduce it with 1 instance and 500 torrents -- might just take longer.  Have you tried that or are you just running 4 instances w/8000 torrents total.

Did you mention your HW?  How much memory is on the machine?  8G?  How many Cores?  What's the %cpu usage during this time... 

Are any of the torrents active?  I.e. actually servicing clients w/traffic?  Are they all waiting for connections, or are some number of them "queued"?   FWIW, I don't use the 'queues' feature of transmission -- so when I have 350 torrents -- they are all awaiting client connections and are all announcing every ~ half hour (I think).  Some of the terminology seems a bit confusing, as when someone says they have 2000 torrents, not sure if all are listening or only a small number are and the rest are queued.

From what I can tell, some people's machines were getting overloaded with more than 10-20 active torrents -- so the queuing was added to allow only some to actually be active at one time.  For myself, I split my torrents when I got about 400, since I noticed my GUI response was getting too slow for my tastes -- but that's running the GUI over a local net via the RPC client.  Just guessing, but probably had to do with the RPC interface and too many requests had to be done to update each display.  Might be helped if the RPC interface had bulk operations, but that presumes anyone can remake/recompile the GUI as I'm told its author has stopped updating it (I tried to compile it, but couldn't get it to compile under the Pascal compiler they used.  

Anwway, when you get time, you might try running 1 instance with 50 clients under Valgrind (it's open source at valgrind.org).  I found it invaluable in tracking down memory problems (mine involved use-after-free pointers).  You wouldn't *want* to run it till it ran out of memory -- would generate way too much output.  Only need to run it through 1 announce (via SSL) cycle, then stop the program -- it will show you any lost memory.  Even 1 client is likely to show the problem if it is really a leak -- even pointer reuse, since it keeps track of everything.

@andreygursky -- BTW , you say 50 torrents with SSL will dup the bug ...  You might try running valgrind as well, even with 1 torrent+SSL and see what it comes up with.  I believe it can find memory problems even with a normal binary -- but to have it mean anything and to track it down, you'd eventually want  to recompile with symbols.

**Comment by Mechazawa** [comment]
I tried @andreygursky's sollution and that seems to have fixed it.

![image](https://user-images.githubusercontent.com/1144171/31009574-744356e4-a4f7-11e7-971a-926ae7e9159e.png)

@Astara 
10Gigs of ram and 4 cores @3.2Gh/z. Transmission-daemon uses barely any CPU at any time. I was able to reproduce it with a single instance and 1k torrents. My torrents are all active and seeding I rarely have more then 3 items queued.

**Comment by nakhan98** [comment]
I can also confirm that compiling transmission 2.92 with libcurl4-openssl-dev and libssl1.0-dev on debian/raspbian appears to fix the memory issues. Thanks @andreygursky !

**Comment by Bisaloo** [comment]
Has someone reported this bug to the debian maintainers?

**Comment by andreygursky** [comment]
@Mechazawa, exactly the spikes I spoke about.

@Bisaloo, I've found this issue here from the Debian's bug report [transmission-daemon 2.92 high memory usage](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=865624).

**Comment by Seeder101** [comment]
so does this issue #333 is relevant ?

**Comment by Seeder101** [comment]
I don't know if Debian is going to fix this downstream?

**Comment by Mechazawa** [comment]
@Seeder101 The package maintainers should be notified: https://packages.debian.org/source/stretch/transmission

**Comment by Seeder101** [comment]
@Mechazawa https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=865624 
this hasn't change since Jun if anyone knows how to interact with this bug/issue in the debian system would be great

**Comment by Astara** [comment]
You mean to "make" the maintainers...???  Huh?  How do you propose to "**make**" a volunteer to do what you want?   Maybe try editing the product generation script (for rpm would be "spec" file.. not sure what it would be for debian)... and submitting the fixed version as a proposed fix?  

*ahem*... and people wonder why developers get burnt out....

**Comment by Seeder101** [comment]
@Astara you got it wrong sir
I didn't mean to force them or something 
and I know it's a volunteer-based...it's like I'm saying kindly if someone can report or let them know about the issue 

and Sorry If I made a false comment or considered bad 

Regards

**Comment by Astara** [comment]
Since others have posted "how to fix the problem" (andreygursky: I've removed libcurl4-gnutls-dev and libssl-dev and installed libcurl4-openssl-dev and libssl1.0-dev instead. Then I've rebuilt transmission. This seems to solve the issue), I'm guessing that the debian folks may include that fix in the next release, but I don't know what debian's policy is in releasing mid-release fixes for individual products when users might be able to apply the same work-around...dunno... don't work w/debian....sorry.. @Seeder101 -- don't work too much about it ... just the wording seemed a bit "amusing" for a sunday...   

I remember a software manager talking about how trying to "manage" software engineers was a bit like trying to herd cats -- how much more difficult it would be to  get volunteer SW engineers to do things... (though depending on the SW engineer, a gift of 6-pack or a case used to help some of them get motivated... ;-))...

**Comment by Seeder101** [comment]
@Astara what do you use for servers?

**Comment by Astara** [comment]
I have a (getting a bit older now), Dell PowerEdge T610 Server @ home that runs linux.  I use win7 for a desktop and use the transmission remote gui to control the transmission daemon.  I used to run 2 xm daemons regularly, but sometimes take 1 down to free up some net bandwidth when I play an occasional multiplayer game.  Running the torrents is usually a background task on the dell -- it also serves as the win7's main internet connect, (proxy, router, DNS, email, backups).   The desktop just died -- so need to get a new one.  The server is still going @ 7+ years.  My previous "server" was a repurposed workstation, also from Dell, that lasted about 10 years.

**Comment by ciprian-radu** [comment]
I tried the solution mentioned here and it works for me. Thank you!

**Comment by andreygursky** [comment]
The main issue with this solution is that it actually a workaround. Transmission should work with an arbitrary SSL library used by curl, since curl fully abstracts the logic. The bug in interaction between curl and curl with ssl library must be more subtle. But since nobody has time to investigate it further using a workaround is a way to go.

@Seeder101,
> if anyone knows how to interact with this bug/issue in the debian system would be great

It's simple, you don't even need to register (as in most modern bug reporting systems), just send an email to the bugnumber@bugs.debian.org, in our case: 865624@bugs.debian.org.

anyone, @nakhan98, @ciprian-radu,
Do you know exactly, whether you have torrents with SSL trackers or only "plain" HTTP (but actually with HTTPS redirection) (UDP is out of interest)? I'd be surprised, that exchanging the SSL library helps for plain text trackers.

**Comment by ciprian-radu** [comment]
@andreygursky I built Transmission from sources using config option `--enable-lightweight` (https://github.com/transmission/transmission/blob/master/configure.ac#L605) as I am running from a Raspberry Pi with 512 MB of RAM. This option has _prefer unencrypted peer connections_ set in Transmission.
Anyway, in my case the usage of Transmission is not that resource intensive but, I do have another service that also takes full CPU and considerable amount of RAM so I suspect when both services were running, my rPI remained without memory.

**Comment by ciprian-radu** [comment]
@Seeder101 Out of curiosity, is Transmission built in Debian with `libcurl4-gnutls-dev` instead of `libcurl4-openssl-dev`? Why, when here https://github.com/transmission/transmission/wiki/Building-Transmission it is clearly specified to use `libcurl4-openssl-dev`.

**Comment by andreygursky** [comment]
@ciprian-radu
> This option has prefer unencrypted peer connections set in Transmission.

I believe it doesn't matter. For me the issue was triggered by SSL tracker connection.

> Out of curiosity, is Transmission built in Debian with libcurl4-gnutls-dev instead of libcurl4-openssl-dev? Why

I guess because openssl is "evil" and should be replaced by gnutls. In respect to license issues I can agree but for transmission it is not a problem.

**Comment by ben-ba** [comment]
Try a build myself (Raspberry Pi), but without success, any hints?

My way (nearly the wiki thing):
```
    mkdir –p ~/builds/transmission/
    wget https://github.com/transmission/transmission-releases/raw/master/transmission-2.92.tar.xz
    tar xf transmission-2.92.tar.xz
    cd transmission-2.92

    CFLAGS="-Os -march=native" ./configure --disable-nls --disable-cli
    make
    checkinstall --fstrans=no
```

gives an timeout after starting transmission-daemon with systemctl

with 
```
    CFLAGS="-Os -march=native" ./configure --disable-nls --disable-cli--with-systemd-daemon
```
there is an error :
```
    checking for SYSTEMD_DAEMON... no
    configure: error: systemd startup notification support requested, but libsystemd-daemon not found.
```

**Comment by Pixa1** [comment]
Hi all,  
i would also like to confirm the problem with transmission on raspbian stretch. Since i'm a noob and didnt know much about static linking that everybody here was talking about, i'll write the procedure to help someone like me.
So basically  @messyidea  pointed me in the right direction and especially with his script i was able to get to the goal:
https://gist.github.com/messyidea/16f9c63d7b219d2a4755e30aedd8268e 

There is a couple of lines in the script that should be modified to work on raspbian, on line 52 it says:
`./Configure linux-x86_64 \ `
change it to
`./Configure linux-generic32 \  `
(please i am no linux expert, so if anyone can correct me on this , please do, but this made it working for me.)
Also on the line 147:
`--disable-nls \`
should be
`--enable-nls \`

when script completes (and it takes a while), there should be /home/pi/transmission/transmission-2.92/daemon directory and in there should be transmission-daemon file, if you check it with file command you should see that it is statically linked:
```
pi@raspberrypi:~/transmission/transmission-2.92/daemon $ file transmission-daemon
transmission-daemon: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, for GNU/Linux 3.2.0, BuildID[sha1]=c89515dde70bd52d07ccf902f91ad7701bd5cb95, not stripped
```

so basically everything you need to do now is to copy this newly compiled transmission-daemon file to /usr/bin/ folder
`pi@raspberrypi:~/transmission/transmission-2.92/daemon $ sudo cp transmission-daemon /usr/bin/`

and Change the /etc/systemd/system/multi-user.target.wants/transmission-daemon.service as @messyidea said (i could only see the difference with type=simple between old service).

Hopefully this will help someone :)

Regards

**Comment by xavery** [comment]
@andreygursky Doing that worked for me as well. Memory usage of the Transmission daemon has been at a constant ~100MB with about 350 mostly idle torrents, as opposed to it reaching almost 3GB previously. Thanks a ton.

**Comment by Esux** [comment]
I know this isn't really the place for 'support', but I've looked just about everywhere trying to work this out.

I tried using the transmission source package to apply the patch, but I can't rebuild the package without installing qt5 etc not something I want to have to install on a headless system.

I have around 6000 torrents seeding on transmission and it's using all 8 gigs of ram and CPU use is pretty high too. So I need a way to apply this workaround without disrupting all my torrents and so forth.

I know it's a long shot, but could someone please help me? Even the test package is using gnutls.

**Comment by Astara** [comment]
Esux wrote:
>
> I tried using the transmission source package to apply the patch, but 
> I can't rebuild the package without installing qt5 etc not something I 
> want to have to install on a headless system.
>
By default, the standard make will try to build everything.

You need to rerun the 'configure' script (try 'configure --help' to
see all options)...there you can make sure you set options you want
and disable or build 'without' ones you don't

Looks like I started with something like:

configure --enable-daemon --without-gtk --enabled-utp 
--without-systemd-daemon --with-inotify

So maybe something like that to start and adjust to what's
optimal for your setup...?

**Comment by Esux** [comment]
I tried that and ran into errors. `error: fakeroot debian/rules clean gave error exit status 2`

So I tried downloading 2.93 and compiling it which worked, but systemd would time out.

`: transmission-daemon.service: Start operation timed out. Terminating.
Failed to start Transmission BitTorrent Daemon.
transmission-daemon.service: Unit entered failed state.
transmission-daemon.service: Failed with result 'timeout'.`

It did work when I ran it manually, but it's a headache.

Forcing me to revert back to the Debian package. Extremely frustrating. I'm moving hosts next month and I sure as hell won't be using Debian again. Guess I'll just need to live with it consuming all my resources until then.

**Comment by sirskills** [comment]
I am a little confused as to the solution. I have upgraded my pi 2 to stretch with the apt-get --dist-upgrade and now have the transmission 2.92 with the memory leak.  (I have a Pi 3B+ that I want to start using but can't because of this bug).

I get that I need to remove libcurl4-gnutls-dev and libssl-dev and installed libcurl4-openssl-dev and libssl1.0-dev instead. Then rebuild transmission but how do I do that?

I can purge the two offending libraries and install the two replacements, but what do I do with the existing transmission? How do I rebuild transmission with the new libraries? 

Sorry if this is a newb question but you folks make it sound so easy.

Thanks.

**Comment by kami83** [comment]
Hi,

i just tested it but compiling with the curl and transmission now i am getting this error:
libtool: warning: library '/root/transmission/opt/lib/libevent.la' was moved.
libtool: warning: library '/root/transmission/opt/lib/libevent.la' was moved.
  CC       blocklist-test.o
  CCLD     blocklist-test
libtool: warning: library '/root/transmission/opt/lib/libevent.la' was moved.
libtool: warning: library '/root/transmission/opt/lib/libevent.la' was moved.
./libtransmission.a(web.o): In function `createEasy':
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:178: undefined reference to `curl_easy_init'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:182: undefined reference to `curl_easy_setopt'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:183: undefined reference to `curl_easy_setopt'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:184: undefined reference to `curl_easy_setopt'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:185: undefined reference to `curl_easy_setopt'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:186: undefined reference to `curl_easy_setopt'
./libtransmission.a(web.o):/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:187: more undefined references to `curl_easy_setopt' follow
./libtransmission.a(web.o): In function `tr_webThreadFunc':
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:384: undefined reference to `curl_global_init'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:385: undefined reference to `curl_global_init'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:407: undefined reference to `curl_multi_init'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:432: undefined reference to `curl_multi_add_handle'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:450: undefined reference to `curl_easy_pause'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:455: undefined reference to `curl_multi_timeout'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:471: undefined reference to `curl_multi_fdset'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:484: undefined reference to `curl_multi_perform'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:496: undefined reference to `curl_easy_getinfo'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:498: undefined reference to `curl_easy_getinfo'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:499: undefined reference to `curl_easy_getinfo'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:500: undefined reference to `curl_easy_getinfo'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:503: undefined reference to `curl_multi_remove_handle'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:505: undefined reference to `curl_easy_cleanup'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:488: undefined reference to `curl_multi_info_read'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:524: undefined reference to `curl_multi_cleanup'
./libtransmission.a(web.o): In function `tr_webGetTaskInfo':
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:549: undefined reference to `curl_easy_getinfo'
./libtransmission.a(web.o): In function `tr_http_unescape':
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:636: undefined reference to `curl_unescape'
/root/transmission/src/transmission/transmission-2.92/libtransmission/web.c:638: undefined reference to `curl_free'
./libtransmission.a(crypto-utils-openssl.o): In function `log_openssl_error':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:42: undefined reference to `ERR_get_error'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:52: undefined reference to `ERR_load_crypto_strings'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:57: undefined reference to `ERR_error_string_n'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_sha1_init':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:101: undefined reference to `EVP_MD_CTX_create'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:103: undefined reference to `EVP_sha1'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:103: undefined reference to `EVP_DigestInit_ex'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:106: undefined reference to `EVP_MD_CTX_destroy'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_sha1_update':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:122: undefined reference to `EVP_DigestUpdate'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_sha1_final':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:137: undefined reference to `EVP_DigestFinal_ex'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:142: undefined reference to `EVP_MD_CTX_destroy'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_rc4_new':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:179: undefined reference to `EVP_CIPHER_CTX_new'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:181: undefined reference to `EVP_rc4'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:181: undefined reference to `EVP_CipherInit_ex'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:184: undefined reference to `EVP_CIPHER_CTX_free'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_rc4_free':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:194: undefined reference to `EVP_CIPHER_CTX_free'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_rc4_set_key':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:205: undefined reference to `EVP_CIPHER_CTX_set_key_length'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:207: undefined reference to `EVP_CipherInit_ex'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_rc4_process':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:226: undefined reference to `EVP_CipherUpdate'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_dh_new':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:239: undefined reference to `DH_new'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:244: undefined reference to `BN_bin2bn'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:245: undefined reference to `BN_bin2bn'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:247: undefined reference to `DH_free'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_dh_free':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:260: undefined reference to `DH_free'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_dh_make_key':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:277: undefined reference to `DH_generate_key'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:280: undefined reference to `BN_bn2bin'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:281: undefined reference to `DH_size'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_dh_agree':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:303: undefined reference to `BN_bin2bn'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:306: undefined reference to `DH_size'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:309: undefined reference to `DH_compute_key'
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:320: undefined reference to `BN_free'
./libtransmission.a(crypto-utils-openssl.o): In function `tr_rand_buffer':
/root/transmission/src/transmission/transmission-2.92/libtransmission/crypto-utils-openssl.c:334: undefined reference to `RAND_bytes'
collect2: error: ld returned 1 exit status
Makefile:1067: die Regel für Ziel „blocklist-test“ scheiterte
make[1]: *** [blocklist-test] Fehler 1
make[1]: Verzeichnis „/root/transmission/src/transmission/transmission-2.92/libtransmission“ wird verlassen
Makefile:507: die Regel für Ziel „all-recursive“ scheiterte
make: *** [all-recursive] Fehler 1

maybe someone can help me ?

cu kami

**Comment by andreygursky** [comment]
@kami83:
> i just tested it but compiling with the curl and transmission now i am getting this error:

Without knowing what OS and steps to rebuild transmission you're using, it is hard to suggest you something definitive.

**Comment by kami83** [comment]
Hi,

i am sorry just forgot. I am using raspbian stretch on a RPI 3B+. I tried your apt-steps, but i have the same problem with the mem. Building works but memory is full.

Any help??
Thanks a lot.

Cu kami

**Comment by ben-ba** [comment]
@kami83 there is an debian/control file.
https://anonscm.debian.org/cgit/collab-maint/transmission.git/tree/debian/control
change line 10 from
` libcurl4-gnutls-dev | libcurl4-dev | libcurl-dev,`
to
`libcurl4-openssl-dev | libcurl-dev,`

**Comment by kami83** [comment]
hi, i just recompile everything under raspbian stretch with my pi 3B+. On this system this is not a openssl/gnutls problem.The memory always gets full and the system break. The problem must be somewhere else. I am no only have installed: libcurl4-openssl-dev libssl1.0-dev. And everything is new but same problem here.

Maybe someone can help.

cu kami

**Comment by xavery** [comment]
In this case, this is not the same issue. Perhaps not even a Transmission-related issue at all.

**Comment by ben-ba** [comment]
@kami i have build transmission successfully this way for the pi 3B, based on transmission_2.92-2+deb9u1_armhf. Give it a try.

Let the originally installed version over apt/apt-get installed.

Save the following binary to `/usr/local/bin/` and change your systemD unit file ( `/etc/systemd/system/multi-user.target.wants/transmission-daemon.service`) this way:

`#ExecStart=/usr/bin/transmission-daemon -f --log-error --logfile /var/log/transmission/transmission-daemon.log
ExecStart=/usr/local/bin/transmission-daemon -f --log-error --logfile /var/log/transmission/transmission-daemon.log`

[https://www.dropbox.com/s/x9aatjdoajsiqry/transmission-daemon?dl=0](url)


PS.: Tonight i can write an step by step guide, which worked for me.

**Comment by andreygursky** [comment]
@sirskills:

> $ apt-get source transmission
>
> Is it just inserting the source from the github page I imagine, like cloning https://github.com/transmission/transmission.git
?

No, apt-get pulls and extracts the source tarball from Debian repository unless you've altered the corresponding configuration file.

> also sudo dpkg -i <transmission packages ...>
the is that all the dependencies?

transmission-daemon_2.93-1_amd64.deb transmission-common_2.93-1_all.deb transmission-remote-cli_1.7.0-1_all.deb (and if needed -gtk, -qt).

**Comment by andreygursky** [comment]
@kami83:

Please post the output of
```
$ dpkg -l | grep -e libcurl -e libssl -e gnutls -e libevent
```

**Comment by sandrotosi** [comment]
folks, it is not that simple you can just swap the packages using gnutls with the one using openssl.

@mikedld as owner of this project, could you read https://people.gnome.org/~markmc/openssl-and-the-gpl.html and add such an exception to transmission license? In the near future, we expect OpenSSL to be relicensed under Apache-2, but we're not there yet.

it's true that the main license for transmission is Expat, but many (if not all) files in libtransmission are under GPL and those are the files also accessing SSL objects from cURL.

once the exception is in place, it would be extra awesome if we could have a transmission release (even a patch version, like 2.74.1) so that Debian can pick it up and switch to `libcurl4-openssl-dev` and hopefully address this bug which many users are so eager to get fixed.

thanks!

**Comment by mikedld** [comment]
@sandrotosi It's already there in [COPYING](/transmission/transmission/blob/master/COPYING):
> In addition, linking to and/or using OpenSSL is allowed.

**Comment by andreygursky** [comment]
@sandrotosi:

> folks, it is not that simple **you** can just swap the packages using gnutls with the one using openssl.

Just to clarify "you": for users it is only a matter of rebuilding packages regardless of license. It's only an issue for those who redistribute the binary packages, i.e. OS distributions.

@mikedld, thanks for the clarification!

**Comment by andreygursky** [comment]
@sandrotosi,

BTW, transmission package has libssl-dev in its build dependencies for over 7 years (and maybe much longer) [1]. So it already links to OpenSSL.

> but many (if not all) files in libtransmission are under GPL and those are the files also accessing SSL objects from cURL.
>
> once the exception is in place, it would be extra awesome if we could have a transmission release (even a patch version, like 2.74.1) so that Debian can pick it up and switch to libcurl4-openssl-dev

Or you investigated this in the past in a more detail per file basis, I mean such files using OpenSSL were not GPL, thus libssl-dev dep? OTOH, you've said "...(if not all) files". In the case of all files this would imply transmission could not link to OpenSSL ever in Debian.

[1] https://salsa.debian.org/debian/transmission/commit/49f786fdc067a4a4ba1313958bd034fd440923b3

**Comment by andreygursky** [comment]
@kami83, looks good besides some mixing:

> ii libgnutls26:armhf 2.12.20-8+**deb7u3** armhf GNU TLS library - runtime library
> ii libssl1.0.0:armhf 1.0.1t-1+**deb8u8** armhf Secure Sockets Layer toolkit - shared libraries

You're already on Debian 9, but you still have packages from 7 and 8. Who knows to what issues this may lead.

It seems, you would have to investigate your issue deeper (ensure all your packages are up-to-date, enable logging, increase verbosity, enable debug logging,..).

**Comment by sirskills** [comment]
Well fingers crossed but it's been running for 12 hours with no issues and I didn't have to do anything since the upgrade from Jessie to stretch on my pi 3B+.

dpkg -l | grep -e libcurl -e libssl -e gnutls -e libevent
ii  libcurl3:armhf                   7.38.0-4+deb8u10                  armhf        easy-to-use client-side URL transfer library (OpenSSL flavour)
ii  libcurl3-gnutls:armhf            7.38.0-4+deb8u10                  armhf        easy-to-use client-side URL transfer library (GnuTLS flavour)
ii  libcurl4-openssl-dev:armhf       7.38.0-4+deb8u10                  armhf        development files and documentation for libcurl (OpenSSL flavour)
ii  libevent-2.0-5:armhf             2.0.21-stable-3                   armhf        Asynchronous event notification library
ii  libgnutls-deb0-28:armhf          3.3.8-6+deb8u7                    armhf        GNU TLS library - main runtime library
ii  libgnutls-openssl27:armhf        3.3.8-6+deb8u7                    armhf        GNU TLS library - OpenSSL wrapper
ii  libssl-dev:armhf                 1.0.1t-1+deb8u8                   armhf        Secure Sockets Layer toolkit - development files
ii  libssl-doc                       1.1.0f-3+deb9u2                   all          Secure Sockets Layer toolkit - development documentation
ii  libssl1.0.0:armhf                1.0.1t-1+deb8u8                   armhf        Secure Sockets Layer toolkit - shared libraries

**Comment by sandrotosi** [comment]
thanks @mikedld i should have a better job at keeping the copyright information up-to-date in debian; 2.94 is being released within minutes linking to libcurl-openssl

**Comment by sirskills** [comment]
Upgraded to my Pi 3B+ to stretch, added the testing repo and installed 2.94. Stable memory usage for three straight days. I'd say problem solved thanks so much everyone.

**Comment by andreygursky** [comment]
@sirskills:

> I'd say problem solved thanks so much everyone.

The problem is solved in Debian since 2.94 packaging [1] by applying the proposed workaround. But in general it is not yet solved. I continue investigating this further.

[1] https://tracker.debian.org/news/955142/accepted-transmission-294-1-source-amd64-all-into-unstable/

**Comment by javicacheiro** [comment]
Running on Pi using a fresh install of stretch with latest version (2018-11-13-raspbian-stretch.zip) I found the same issue. The memory consumption steadidly increased until the OOM  Killer stopped the process.

Upgrading to the 2.94 solves the issue, unfortunately this involves installing it from to the debian testing repo and mixing packages from stable and testing versions.

**Comment by dkimmortal** [comment]
@javicacheiro can you give a walkthrough on how to install 2.94 on arm device? How to mix and match etc...if possible thanks

**Comment by javicacheiro** [comment]
@dkimmortal this is the process I followed:
- Activate the raspbian testing repo adding the following line to /etc/apt/sources.list
```
# Testing branch to get transmission-daemon 2.94 and solve memory leak in 2.92
deb http://raspbian.raspberrypi.org/raspbian/ testing main contrib non-free rpi
```
- Update the cache of packages
```
apt update
```
- Install latest version of transmission-daemon
```
apt install transmission-daemon
```
For me this has installed transmission-daemon version 2.94-1+b2, but it could change with future updates of the repo.
To force this specific version you can request it explicitly, but in general I would stick with the latest one.
```
apt install transmission-daemon=2.94-1+b2
```

Afterwards you can disable the testing repo and continue in a modified stretch system or do the complete jump to the testing version running an `apt dist-upgrade` (but this is not needed to get transmission-daemon 2.94 working).

Main problem is that, since you mix packages from stable and testing versions, this could cause some conflicts when installing additional software mainly because of incompatible library versions.

**Comment by andreygursky** [comment]
@javicacheiro:
> ```# Testing branch to get transmission-daemon 2.94 and solve memory leak in 2.92```

Just to avoid further confusions: it is not a memory leak but behavior that can be called "excessive memory allocation". Many not small enough memory chunks allocated almost at once, then being freed on regular basis (or at high rate).

It is provoked by suboptimal curl build configuration options how to access SSL certificates data, suboptimal code in gnutls to access this data and how curl uses this code. OpenSSL turned out to be much more efficiently managing SSL verification. I guess, not much one could fix in transmission itself.

**Comment by javicacheiro** [comment]
@andreygursky thanks a lot for the clarification. I have updated the comment to avoid confusion.
If I understand it correctly it would be possible to fix the problem in 2.92 recompiling the package using OpenSSL instead of gnutls as @nakhan98 has done. I see there were licensing concerns but since 2.94 in the testing repo is already compiled using gnutls, I guess they have been solved. So it should be possible to do the same with 2.92 in the stable repo. Do you know what could be blocking this fix to go into the stable repo?

**Comment by MichaIng** [comment]
@andreygursky 
Perhaps there is just a push/request required. The initial bug report has been fixed and closed, so possibly it is simply out of scope for the maintainer(s): https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=865624

If there is not already an opened bug report/request for this, I will see if I add a comment to the bug above to ask for backport.

**Comment by admusin** [comment]
@andreygursky thanks a bunch for the details provided.
I'd just recently upgraded my NAS to a Debian Stretch-based version of its OS (OMV) and this issue was driving me nuts.
After recompiling transmission 2.92, yesterday, with the libraries you mentioned the issue didn't appear anymore, so far.

**Comment by sergey-alekseev** [comment]
Unfortunately, upgrading to 2.94 didn't solve the problem for me.

**Comment by andrasveto** [comment]
I continue to have this issue as well, on Ubuntu 16.04.6 LTS with Transmission 2.94. http announce urls work perfectly, but https invariably results in a memory leak that takes up all available memory within hours.

**Comment by MichaIng** [comment]
@sergey-alekseev @andrasveto 
Not sure if it's the upstream version that fixes or even produces this error. On Debian, as far as can be seen from the bug report, the high memory usage was produced by having it compiled with `libcurl4-gnutls-dev`, which does not match the build instructions of this Wiki as well, and was resolved by compiling with `libcurl4-openssl-dev`, which matches this Wiki build instructions.

Perhaps Ubuntu as well and still builds with the wrong library, regardless of upstream version update?

**Comment by sergey-alekseev** [comment]
@MichaIng thanks for the details. I followed "Build instructions: https://github.com/transmission/transmission/wiki/Building-Transmission#on-unix" on my Raspbian. However I got an error during building transmission:  
```
checking for LIBEVENT... configure: error: Package requirements (libevent >= 2.0.10) were not met:

No package 'libevent' found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables LIBEVENT_CFLAGS
and LIBEVENT_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
```

Will try to resolve the problem later and will report here in several days whether I get or don't get any problems.

**Comment by MichaIng** [comment]
@sergey-alekseev 
If you compiled libevent according to the instructions, I guess the custom compilation creates it in /usr/local/lib which is probably not part of default path where the transmission compilation then looks for it. Perhaps `ldconfig` can solve it.

But if you are on Raspbian Buster, there is a more current version available, that should work, as the build instructions say "get the latest version" and the error message "libevent >= 2.0.10". On Buster `apt install libevent-dev` installs v2.1.8 which should thus work.

**Comment by sergey-alekseev** [comment]
@MichaIng thanks for your help.

Yes, I'm on Raspbian Buster which I had to install first to be able to install transmission 2.94.  

Now I get another problem which I'll try to resolve later (failing `mkdir -p` for some reason):  
```
 /bin/mkdir -p '/usr/local/share/transmission/web/images'
/bin/mkdir: cannot create directory '/usr/local/share/transmission': No such file or directory
make[3]: *** [Makefile:408: install-dist_dataDATA] Error 1
make[3]: Leaving directory '/var/tmp/transmission-2.94/web/images'
make[2]: *** [Makefile:477: install-am] Error 2
make[2]: Leaving directory '/var/tmp/transmission-2.94/web/images'
make[1]: *** [Makefile:499: install-recursive] Error 1
make[1]: Leaving directory '/var/tmp/transmission-2.94/web'
make: *** [Makefile:510: install-recursive] Error 1

****  Installation failed. Aborting package creation.
```

Will be posting my progress here to help other people.

**Comment by MichaIng** [comment]
@sergey-alekseev
That is indeed strange since `-p` is exactly to resolve this error by pre-creating all required parent dirs. Does it fail as well when you call this command manually?

**Comment by sergey-alekseev** [comment]
@MichaIng  
>Does it fail as well when you call this command manually?

No, it succeeds.  

I succeeded after several times manually creating necessary directories after similar errors (`No such file or directory`).  

I managed to get it working after some problems, then deleted `/etc/transmission-daemon/settings.json` because it seemed after building it the config file became `.config/transmission-daemon/settings.json`. And now it's not working again... So one more try probably later...  

```
$ curl http://192.168.0.2:9091
curl: (7) Failed to connect to 192.168.0.2 port 9091: Connection refused
```

Running `sudo pkill -HUP transmission-da` and `sudo service transmission-daemon restart` doesn't help while having both `/etc/transmission-daemon/settings.json` and `.config/transmission-daemon/settings.json` now.

**Comment by MichaIng** [comment]
@sergey-alekseev 
Note that it depends on the command line option in the first place, and the user in the second place, which config file is used. When you run from terminal with some user, `~/.config/transmission-daemon/settings.json` will be used, or the config dir created, if it does not exist.
Or one can use the command line option `-g /path/to/config/dir` to specify a specific directory.

The default Debian package e.g. comes by default with `/etc/transmission-daemon/settings.json`. But actually the `debian-transmission` has its home dir set to `/var/lib/transmission-daemon` which contains the settings file `/var/lib/transmission-daemon/.config/transmission-daemon/settings.json` which is a symlink to `/etc/transmission-daemon/settings.json`. So a bit confusing loop to satisfy some standards like run user home dirs in `/var/lib/<name>` and daemon config files in `/etc/<name>` with the given possibilities of the `transmission-daemon` binary.
Long story short, `/etc/transmission-daemon` is not used when you do a custom build, so you can remove it.

Which user runs the daemon (`systemctl cat transmission-daemon`) and did you verify that you created/edited the correct config file in its home dir (`getent passwd transmission-daemon`)?
_____________
I am still confused about the `mkdir -p` error, especially if you ran both, `make` and `mkdir` (manually to pre-create the dirs), with the same user. I though that probably this error would show up, if permissions are missing, but actually just tried and it then says so instead of "no such dir"... Probably some environment black magic that I do not understand 🤣, however, glad you managed to finish the build.

**Comment by sergey-alekseev** [comment]
@MichaIng thanks for the details. However no idea at the moment why it's not working.

```
sergey@raspberrypi:~ $ systemctl cat transmission-daemon
# /run/systemd/generator.late/transmission-daemon.service
# Automatically generated by systemd-sysv-generator

[Unit]
Documentation=man:systemd-sysv-generator(8)
SourcePath=/etc/init.d/transmission-daemon
Description=LSB: Start or stop the transmission-daemon.
Before=multi-user.target
Before=multi-user.target
Before=multi-user.target
Before=graphical.target
After=remote-fs.target
After=network-online.target
Wants=network-online.target

[Service]
Type=forking
Restart=no
TimeoutSec=5min
IgnoreSIGPIPE=no
KillMode=process
GuessMainPID=no
RemainAfterExit=yes
SuccessExitStatus=5 6
ExecStart=/etc/init.d/transmission-daemon start
ExecStop=/etc/init.d/transmission-daemon stop
ExecReload=/etc/init.d/transmission-daemon reload
```

I run transmission daemon as `sudo service transmission-daemon start`. I have both `/etc/transmission-daemon/settings.json` and `~/.config/transmission-daemon/settings.json`.

---

I'm confused about the `mkdir -p` error as well.

**Comment by MichaIng** [comment]
@sergey-alekseev 
Ah it's a sysvinit service auto-translated into a systemd unit. Isn't there some compiling option to create a systemd unit on install?

At least a unit is available:
```
wget https://raw.githubusercontent.com/transmission/transmission/develop/daemon/transmission-daemon.service -O /etc/systemd/system/transmission-daemon.service
systemctl daemon-reload
systemctl enable transmission-daemon
```

However not required, sysvinit should do as well. To check the run user then:
```
cat /etc/init.d/ transmission-daemon
```
Somewhere around the first lines bust be a USER entry. The following then can be used to find the users home dir, thus get the config file location that is used.
```
getent passwd <username>
```

**Comment by sergey-alekseev** [comment]
Thanks for your help @MichaIng. Unfortunately, I haven't succeeded building Transmission. So I switched back to using `transmission-daemon` installed via `apt-get`. The issue is annoying, but not that much as not having working torrents. Hopefully someone will succeed building Transmission and getting rid of the memory issue following the comments.

**Comment by Lipown** [comment]
I spend a night over this problem without the solution. Running Raspbain Buster on Pi 4 4gb.

**Comment by Mechazawa** [comment]
> I spend a night over this problem without the solution. Running Raspbain Buster on Pi 4 4gb.

Have you tried @andreygursky's suggestion? That did the trick for me

**Comment by Lipown** [comment]
> > I spend a night over this problem without the solution. Running Raspbain Buster on Pi 4 4gb.
> 
> Have you tried @andreygursky's suggestion? That did the trick for me

I tried @javicacheiro  solution as it should be solved in the newest release. But it is unfortunatelly not. As I am quite noob, the recompilation seems to be comlicated for me.

**Comment by Mechazawa** [comment]
```
# Prepapare
sudo apt remove libcurl4-gnutls-dev libssl-dev
sudo apt install libcurl4-openssl-dev libssl1.0-dev git wget gcc cmake

# Download Transmission
wget https://github.com/transmission/transmission/archive/2.94.tar.gz
tar xf transmission-2.94.tar.gz
rm transmission-2.94.tar.gz
cd transmission-2.94

# Build
mkdir build
cd build
cmake ..
make

# Install
sudo apt remove transmission-daemon
sudo make install
```

I haven't tried it yet but this script should
 - uninstall transmission
 - uninstall the bad SSL dependencies
 - install the good SSL dependencies
 - install (some) build tools for compiling
 - download transmission 2.94
 - build transmission 2.94
 - install transmission 2.94 (fixed without the memory leak)

It might be that it complains that it can't find a dependency during the build process but that should be fixed by installing said dependency.

**Comment by Lipown** [comment]
> ```
> # Prepapare
> sudo apt remove libcurl4-gnutls-dev libssl-dev
> sudo apt install libcurl4-openssl-dev libssl1.0-dev wget gcc cmake
> 
> # Download Transmission
> wget https://github.com/transmission/transmission/archive/2.94.tar.gz
> tar xf transmission-2.94.tar.gz
> rm transmission-2.94.tar.gz
> cd transmission-2.94
> 
> # Build
> mkdir build
> cd build
> cmake ..
> make
> 
> # Install
> sudo apt remove transmission-daemon
> sudo make install
> ```
> 
> @Lipown I haven't tried it yet but this script should
> 
> * uninstall the bad SSL dependencies
> * install the good SSL dependencies
> * install (some) build tools for compiling
> * download transmission 2.94
> * build transmission 2.94
> * uninstall transmission
> * install transmission 2.94 (fixed without the memory leak)
> 
> It might be that it complains that it can't find a dependency during the build process but that should be fixed by installing said dependency.

Thanks. I just get errors executing second line. The output is (sorry its auto translated from czech):


795/5000
Loading package lists… Done
A dependency tree is created
Loading status information… Done
gcc is the latest version (4: 8.3.0-1 + rpi2).
gcc is set to be installed manually.
git is the latest version (1: 2.20.1-2 + deb10u1).
wget is already the latest version (1.20.1-1.1).
wget set to be installed manually.
Some packages could not be installed. This may mean that you require
an impossible situation, or if you are using an unstable distribution of that
required packages have not yet been created or moved from Incoming Queue.
The following information can help you solve this situation:

The following packages have unfulfilled dependencies:
  libcurl4-openssl-dev: Consolidates with: libssl1.0-dev but 1.0.2q-2 will install
E: Problems cannot be fixed, some packages are in a broken state.

**Comment by Lipown** [comment]
Nope, I installed one package by one without error, but in compliation and after make, there are many errors unfortunatelly

**Comment by colethegamer123** [comment]
Same problem. Solved for months, now came back again.

Debian 10. Transmission 3.00

**Comment by sirskills** [comment]
The issue came back again, Linux raspberrypi 5.4.72-v7l+ Transmission-daemon 2.9.4. At least this time it is very slow, takes a several days for it to max out my 1GB RAM, starting from idle at around 250MB used, on the Pi and luckily is just kills the process and not my pi, so I have to restart transmission.

**Comment by afk11** [comment]
I've noticed oom-reaper killing transmission daemon a few times a week.. I put on prometheus' nodeexporter utility to monitor memory usage.. It seems to grow at 13.6 Mb per minute until it crashes (458Mb to 737Mb over 20 minutes) now. I'll monitor and see if this changes as some finish. I've added 'Restart=always' to the systemd configuration in the hope it makes the problem less obvious, but 

debian9, kvm guest, transmission-daemon 2.92 (14714)

**Comment by andreygursky** [comment]
> Next, libraries used by transmission-daemon, plus the debian package, it's version, and the filesystem location.
> 
> libcurl3-gnutls Version: 7.52.1-5+deb9u12 /usr/lib/x86_64-linux-gnu/libcurl-gnutls.so.4

This is what makes trouble. transmission should be built with libcurl4-openssl-dev installed (instead of libcurl4-gnutls-dev).

**Comment by afk11** [comment]
@andreygursky I guess I've been lucky this hasn't affected my for very long, but for me, this issue only started over the last 3-4 months. I edited my post above, but on 2020-07-30 I installed version 7.52.1-5+deb9u10 of libcurl3-gnutls. I didn't have the issue back when I setup the box - the version previous was installed 2020-02-26 and was version 7.52.1-5+deb9u9

**Comment by andreygursky** [comment]
> I looked up when the package was updated on my system.

Downgrading is of course the simplest solution. But Debian oldstable packages are likely to be updated only because of security issues, thus making downgrades only for short time justifiable.

**Comment by afk11** [comment]
@andreygursky it's not crashing fast enough to make it unusable, it just took 1 hour to get killed. I added 'Restart=always' to the systemd config so I don't have to intervene.

I'll wait 20 minutes, see how fast memory increases, then see if deleting some torrents helps.

**Comment by andreygursky** [comment]
> if deleting some torrents helps

Instead it is enough to ensure, that not too many torrents with http**s** trackers are announced almost simultaneously. You may try to stop them and then start one after another with some delay.

**Comment by afk11** [comment]
I leave my torrents paused when finished (for shame) but, but could transmission be connecting to the trackers even when the torrent is paused?

I removed all torrents, and it seemed to stop growing, but stuck at 460Mb (total used ram on system). I didn't wait for long, but restarting it dropped it back down to 109Mb used by the system and it doesn't seem to be growing, so yay, problem solved.

I guess it's a deterant to seeding if you feel you should remove them all after 50 or something :/ but if this makes the crashes go away, that's what I'm going to do

**Comment by andreygursky** [comment]
> but could transmission be connecting to the trackers even when the torrent is paused?

No, it shouldn't.

**Comment by Lipown** [comment]
Try this https://github.com/raspberrypi/linux/issues/3210#issuecomment-716510916 - I did not tried it on transission but in my case works so far well.

**Comment by sirskills** [comment]
Mine has been stable for 2+ days now. I think my issue is actually coming from the transgui on my windows machine. If I leave it open it seems to chew into the memory on my pi's daemon. I use SSL for the transgui.

**Comment by afk11** [comment]
@sirskills Actually I can confirm this sequence of events.. Mine had been running smoothly but a crash happened overnight. I wasn't downloading anything yesterday, but I did open up transmission-remote-gtk and the connections to the server are using TLS (terminated by nginx though, not transmission-daemon) There was no traffic to the machine yesterday except from 9pm, and that's the moment memory started to rise. I don't have a good explanation for why the memory-usage rate is different, but maybe my other machine has it open too.

**Comment by afk11** [comment]
![transmission-remote-crash-cropped](https://user-images.githubusercontent.com/5617245/98825058-f656a800-242b-11eb-9cf8-6f13b8de30c1.png)

**Comment by sirskills** [comment]
> 
> 
> @sirskills pretty sure I can confirm this sequence of events.. Mine had been running smoothly but a crash happened overnight. I wasn't downloading anything yesterday, but I did open up transmission-remote-gtk and the connections to the server are using TLS (terminated by nginx though, not transmission-daemon)
> 
> There was no traffic to the machine yesterday except from 9.11pm, and a little afterwards traffic started to rise. I don't think I did anything with the downloads around then..

I am also using nginx reverse proxy, could be something.

**Comment by afk11** [comment]
I'll try take it a step further and get memory usage of nginx + transmission-daemon over time, since we're just seeing kernel vs userspace above

**Comment by afk11** [comment]
Oh, but I've only seen oom-killer reap the transmission-daemon process - not nginx. So we need to see what part of transmission-daemon's memory is growing.. maybe pmap will help

**Comment by sirskills** [comment]
> 
> 
> Oh, but I've only seen oom-killer reap the transmission-daemon process - not nginx. So we need to see what part of transmission-daemon's memory is growing..
> 
> Going to log the pid's smap file for a few hours and see what's happening https://unix.stackexchange.com/questions/36450/how-can-i-find-a-memory-leak-of-a-running-process

Yea I haven't seen memory increase on nginx, just the transmission-daemon. Not sure what pmap is.

**Comment by afk11** [comment]
I might have to recompile transmission-daemon to enable all the debug symbols, no line numbers currently, but here's a report by the memleax tool after attaching to my running transmission-daemon https://gist.github.com/afk11/f97b25952016195f9944e8cee325a857

**Comment by afk11** [comment]
I ran strings on the memory region that grew (according to pmap), the start has torrent filenames, some tracker urls, etc, but soon it moves onto really random strings.

At the very end of the file, it's complete gibberish
Just before it turns completely gibberish, I see sections like this:
running strings on the memory dump from GDB:
```
// output trimmed
?dAA*
AplI0.
UUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUU
8 AE
8&d 
// output trimmed
```

and if we scroll around a bit more, I noticed some batches of UUUUU were prefixed by LAME3.9something

```
// cat dump-outputfile.dump | grep LAME
LAME3.98
$LAME3.98UUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUU
LAME3.98
LAME3.98
*LAME3.98
LAME3.98
LAME3.93UUUUUU
```
Probably not important, will try recompile later

**Comment by sirskills** [comment]
Mine has been stable for at least 10 days now. All I've had to do is make sure I don't leave my transmission-remote GUI open on my windows machine.

**Comment by Tharn** [comment]
I experienced this issue even with the "fixed" Debian version 2.94-2 and openssl in use instead of gnutls. It wasn't as pronounced as some people here experienced it, but transmission-daemon still grew over 2-3 days from 45MB (same size as Windows with same torrents) to 180MB.

With transmission-daemon 3.0 backported I am no longer seeing it. Initial memory was again around 45MB, and then it grew pretty quickly to 55MB as torrents were added and downloaded. But so far it's been relatively stable at 55MB.

This is with around 650 seeded torrents, btw. If memory use changes noticeably one way or the other, I will edit.

**Comment by ckerr** [comment]
I'm closing this issue due to inactivity and due to the last report that this is resolved by using 3.0

**Comment by hasezoey** [comment]
i think i have a similar problem on ubuntu 22.04 with transmission-daemon, with just 6 torrent files running, it feels like it keeps all the torrents directly in memory in the process itself instead of using streaming or linux's build-in caching:

```txt
$ ps -aux | grep transmission
debian-+    1711  0.8 39.7 10632612 6461816 ?    Ssl  Apr30 133:43 /usr/bin/transmission-daemon -f --log-error
```

![htop_sylvi](https://user-images.githubusercontent.com/10911626/167877936-eee0cff3-b569-447d-b88c-121e971b72b5.png)

```txt
$ apt list --installed "transmission*"
Listing... Done
transmission-cli/jammy,now 3.00-2ubuntu2 amd64 [installed]
transmission-common/jammy,now 3.00-2ubuntu2 all [installed,automatic]
transmission-daemon/jammy,now 3.00-2ubuntu2 amd64 [installed]
```

Also, when i add another torrent (that is about ~2GB), it over the time of the torrent downloading increases by about another GB of memory usage

**Comment by stevenengler** [comment]
> i think i have a similar problem on ubuntu 22.04 with transmission-daemon, with just 6 torrent files running, it feels like it keeps all the torrents directly in memory in the process itself instead of using streaming or linux's build-in caching:

I've also run into memory issues since upgrading to Ubuntu 22.04. Over about 3 days, the memory usage of transmission-daemon climbs until all memory is used and the entire system (a rpi) freezes. This is with no active torrents. I did not have this issue on Ubuntu 20.04.

**Comment by andreygursky** [comment]
It looks like a different issue. In few words, this issue is about very high memory consumption spikes during reannouncing torrents to trackers using https (instead of plain http), especially if high number of such reannounces are scheduled simultaneously. The workaround is to build transmission with libcurl4-openssl-dev instead of libcurl4-gnutls-dev, which was applied in Debian/Ubuntu. @hasezoey, could you please open a new issue with the description you provided? Then @stevenengler could copy the comment there.
Thanks.

**Comment by hasezoey** [comment]
> It looks like a different issue. In few words, this issue is about very high memory consumption spikes during reannouncing torrents to trackers using https (instead of plain http), especially if high number of such reannounces are scheduled simultaneously. The workaround is to build transmission with libcurl4-openssl-dev instead of libcurl4-gnutls-dev, which was applied in Debian/Ubuntu. @hasezoey, could you please open a new issue with the description you provided? Then @stevenengler could copy the comment there. Thanks.

created a new issue https://github.com/transmission/transmission/issues/3077

**Comment by andreygursky** [comment]
> I think there is a garbage collection missing after spikes.

This issue is not about GC (transmission doesn't use GC). Your issue seems to be a different one: see #3055 (and #3077).
